From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mykyta Komarnytskyy <nkomarn@hotmail.com>
Date: Sat, 24 Oct 2020 21:03:53 -0500
Subject: [PATCH] (Yatopia) Smarter statistics ticking

In vanilla, statistics that count time spent for an action (i.e. time played or sneak time) are incremented every tick. This is retarded. With this patch and a configured interval of 20, the statistics are only ticked every 20th tick and are incremented by 20 ticks at a time. This means a lot less ticking with the same accurate counting.
With an interval of 20, this patch saves roughly 3ms per tick on a server w/ 80 players online.

Original code by YatopiaMC, licensed under MIT
You can find the original code on https://github.com/YatopiaMC/Yatopia

// Sustopia - hardcode interval to 20, mimic Mirai

diff --git a/src/main/java/net/minecraft/world/entity/player/EntityHuman.java b/src/main/java/net/minecraft/world/entity/player/EntityHuman.java
index 5f9e64df007ebc40f7bcb50be495b10e51d5b87a..965ea0da287009c82c434796ee6d491d07f48208 100644
--- a/src/main/java/net/minecraft/world/entity/player/EntityHuman.java
+++ b/src/main/java/net/minecraft/world/entity/player/EntityHuman.java
@@ -265,18 +265,22 @@ public abstract class EntityHuman extends EntityLiving {
         this.p();
         if (!this.world.isClientSide) {
             this.foodData.a(this);
-            this.a(StatisticList.PLAY_ONE_MINUTE);
+            // Yatopia start
+            if (ticksLived % 20 == 0) {
+            this.a(StatisticList.PLAY_ONE_MINUTE, 20);
+            // Yatopia end
             if (this.isAlive()) {
-                this.a(StatisticList.TIME_SINCE_DEATH);
+                this.a(StatisticList.TIME_SINCE_DEATH, 20); // Yatopia // Sustopia
             }
 
             if (this.bx()) {
-                this.a(StatisticList.SNEAK_TIME);
+                this.a(StatisticList.SNEAK_TIME, 20); // Yatopia // Sustopia
             }
 
             if (!this.isSleeping()) {
-                this.a(StatisticList.TIME_SINCE_REST);
+                this.a(StatisticList.TIME_SINCE_REST, 20); // Yatopia // Sustopia
             }
+            } // Yatopia
         }
 
         int i = 29999999;
